File: rustdroid/app/src/main/java/com/example/rustdroid/Manifest.kt
package com.example.rustdroid

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
data class Mpd(@SerialName("media_presentation_duration") val mediaPresentationDuration: String)

@Serializable
data class Representation(
        val id: String,
        val bandwidth: Int,
        val codecs: String,
        val height: String,
        val width: String
)

@Serializable
data class SegmentTemplate(
        val initialization: String,
        @SerialName("start_number") val startNumber: Int,
        val media: String,
        val timescale: Int,
        val duration: String
)

@Serializable
data class Manifest(
        val mpd: Mpd,
        @SerialName("base_url") val baseUrl: String,
        val representations: List<Representation>,
        @SerialName("segment_template") val segmentTemplate: SegmentTemplate,
        @SerialName("segment_count") val segmentCount: Int
) {
    companion object {
        private val json = Json { ignoreUnknownKeys = true }

        fun get(baseUrl: String): Manifest? {
            val jsonString = PlustCore.getManifestJson(baseUrl) ?: return null
            return try {
                json.decodeFromString<Manifest>(jsonString)
            } catch (e: Exception) {
                null
            }
        }
    }
}

File: swust/swust/Manifest.swift
import Foundation

struct Mpd: Codable {
	let mediaPresentationDuration: String

	enum CodingKeys: String, CodingKey {
		case mediaPresentationDuration = "media_presentation_duration"
	}
}

struct Representation: Codable {
	let id: String
	let bandwidth: Int
	let codecs: String
	let height: String
	let width: String
}

struct SegmentTemplate: Codable {
	let initialization: String
	let startNumber: Int
	let media: String
	let timescale: Int
	let duration: String

	enum CodingKeys: String, CodingKey {
		case initialization
		case startNumber = "start_number"
		case media
		case timescale
		case duration
	}
}

struct Manifest: Codable {
	let mpd: Mpd
	let baseUrl: String
	let representations: [Representation]
	let segmentTemplate: SegmentTemplate
	let segmentCount: Int

	enum CodingKeys: String, CodingKey {
		case mpd
		case baseUrl = "base_url"
		case representations
		case segmentTemplate = "segment_template"
		case segmentCount = "segment_count"
	}

	static func get(baseUrl: String) throws -> Manifest {
		guard let jsonPtr = get_manifest_json(baseUrl) else {
			throw NSError(
				domain: "Manifest",
				code: 1,
				userInfo: [NSLocalizedDescriptionKey: "Failed to get manifest"]
			)
		}
		defer { free_string(UnsafeMutablePointer(mutating: jsonPtr)) }
		let json = String(cString: jsonPtr)
		return try JSONDecoder().decode(Manifest.self, from: Data(json.utf8))
	}
}
